core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -322,6 +310,9 @@ private void removeUnallocatedIndices(final ClusterState state) {
         for (AllocatedIndex<? extends Shard> indexService : indicesService) {
             Index index = indexService.index();
             if (indicesWithShards.contains(index) == false) {
+                assert state.metaData().index(index) != null || event.isNewCluster() :
+                    "if the index does not exist in the cluster state, it should either " +
+                    "have been deleted or the cluster must be new";
```

Comment:
can you explain in a comment here why it's ok to remove the index instead of deleting it when `event.isNewCluster()` (i.e. dangling indices)?
Can you also add `index` to the assertion error message.